Use GNUInstallDirs and add missing .pc files for cmake build

dracwyrm opened the pull request:
Use GNUInstallDirs to allow a standard convention in install
variable naming.

Add missing OpenEXR.pc, OpenEXR_Viewers.pc, and PyIlmBase.pc files:
The Autotools build system installs a .pc file, but CMake doesn't.
This adds the missing file but makes use of GNUInstallDirs for
the paths. It is now modeled after the Autotools version.
Based on: PR #138 but corrected to link to the correct libraries.

Changed the .pc files to be external templates like autotools.

Signed-off by: Jonathan Scruggs (j.scruggs@gmail.com)

meshula commented:
LGTM

dracwyrm commented:
I can't remember, but I think I left some hard coded paths in place as I didn't know how they would react on other systems. I'll take another look through to see if I did and report back tomorrow.

dracwyrm commented:
Found a bug in the .pc file. Forgot to add in Requires: IlmBase, so it causes Gimp to fail. I'll be updating this PR.

dracwyrm commented:
Just pushed a final round of changes.

LGTM?

dracwyrm commented:
I changed the .pc files to be external templates instead of writing them. OpenEXR needs to depend on pkg-config being installed to generate the files. If it's not installed, then there's no point in installing them.

Also, can someone confirm that PyIlmBase and OpenEXR_Viewers need .pc files? The autotools builder of those install them, but not the CMake build system. Should I have CMake install them as well?

Thanks.

dracwyrm commented:
@lgritz Another round of GNUInstallDirs goodness. :D

dracwyrm commented:
I updated the patch and rebased it on current develop. @lgritz LGTM?

lgritz commented:
LGTM, though I hope that before a merge, some people can try it out on a few different platform configuration to verify that all is ok in practice.

dracwyrm commented:
The majority of the code is a straight to GNUInstallDirs variable conversion, but the other bit are some fixes as described in the rewritten original post. :) Having people try it would be good. I'll see if @SoapGentoo can look over the file.

SoapGentoo commented on IlmBase/CMakeLists.txt:
```diff
@@ -207,25 +212,11 @@ ADD_SUBDIRECTORY ( ImathTest )
-  lib/pkgconfig
-)
+if ( PKG_CONFIG_FOUND )
+	configure_file(IlmBase.pc.cmake "${PROJECT_BINARY_DIR}/IlmBase.pc" @ONLY)
```
`PROJECT_BINARY_DIR` -> `CMAKE_CURRENT_BINARY_DIR`

SoapGentoo commented on IlmBase/CMakeLists.txt:
```diff
@@ -207,25 +212,11 @@ ADD_SUBDIRECTORY ( ImathTest )
-)
+if ( PKG_CONFIG_FOUND )
+	configure_file(IlmBase.pc.cmake "${PROJECT_BINARY_DIR}/IlmBase.pc" @ONLY)
+	install( FILES "${PROJECT_BINARY_DIR}/IlmBase.pc" DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig COMPONENT dev )
```
same here

SoapGentoo commented on IlmBase/IlmBase.pc.cmake:
```diff
@@ -0,0 +1,9 @@
+libdir=@CMAKE_INSTALL_FULL_LIBDIR@
+includedir=@CMAKE_INSTALL_FULL_INCLUDEDIR@
+OpenEXR_includedir=@CMAKE_INSTALL_FULL_INCLUDEDIR@/OpenEXR
```
I would just inline this

SoapGentoo commented on OpenEXR/CMakeLists.txt:
```diff
@@ -25,8 +25,8 @@ set(CPACK_SOURCE_PACKAGE_FILE_NAME
 INCLUDE ( CPack )
-
-
+INCLUDE ( GNUInstallDirs )
```
lowercase

SoapGentoo commented on OpenEXR/CMakeLists.txt:
```diff
@@ -57,8 +57,8 @@ IF (NOT WIN32)
-INCLUDE_DIRECTORIES ( ${ILMBASE_PACKAGE_PREFIX}/include/OpenEXR )
-LINK_DIRECTORIES ( ${ILMBASE_PACKAGE_PREFIX}/lib )
+INCLUDE_DIRECTORIES ( ${ILMBASE_PACKAGE_PREFIX}/${CMAKE_INSTALL_INCLUDEDIR}/OpenEXR )
+LINK_DIRECTORIES ( ${ILMBASE_PACKAGE_PREFIX}/${CMAKE_INSTALL_LIBDIR} )
```
this should rather be handled by pkgconfig, and only then with an explicit fallback

SoapGentoo commented on OpenEXR/CMakeLists.txt:
```diff
@@ -256,5 +256,10 @@ INSTALL ( FILES
   )
+
+if ( PKG_CONFIG_FOUND )
+	configure_file(OpenEXR.pc.cmake "${PROJECT_BINARY_DIR}/OpenEXR.pc" @ONLY)
```
`PROJECT_BINARY_DIR` -> `CMAKE_CURRENT_BINARY_DIR`

SoapGentoo commented on OpenEXR/CMakeLists.txt:
```diff
@@ -256,5 +256,10 @@ INSTALL ( FILES
+
+if ( PKG_CONFIG_FOUND )
+	configure_file(OpenEXR.pc.cmake "${PROJECT_BINARY_DIR}/OpenEXR.pc" @ONLY)
+	install( FILES "${PROJECT_BINARY_DIR}/OpenEXR.pc" DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig COMPONENT dev )
```
ditto

SoapGentoo commented on OpenEXR/OpenEXR.pc.cmake:
```diff
@@ -0,0 +1,12 @@
+libdir=@CMAKE_INSTALL_FULL_LIBDIR@
+includedir=@CMAKE_INSTALL_FULL_INCLUDEDIR@
+OpenEXR_includedir=@CMAKE_INSTALL_FULL_INCLUDEDIR@/OpenEXR
```
inline

SoapGentoo commented on PyIlmBase/CMakeLists.txt:
```diff
@@ -36,25 +37,36 @@ SET (BUILD_WITH_INSTALL_RPATH 1)
 
 INCLUDE_DIRECTORIES (
-	${ILMBASE_PACKAGE_PREFIX}/include/OpenEXR
+	${CMAKE_INSTALL_FULL_INCLUDEDIR}/OpenEXR
```
again, will likely need pkgconfig

SoapGentoo commented on PyIlmBase/CMakeLists.txt:
```diff
@@ -36,25 +37,36 @@ SET (BUILD_WITH_INSTALL_RPATH 1)
 )
 
-LINK_DIRECTORIES ( ${ILMBASE_PACKAGE_PREFIX}/lib )
+LINK_DIRECTORIES ( ${CMAKE_INSTALL_FULL_LIBDIR} )
```
same

SoapGentoo commented on PyIlmBase/PyIlmBase.pc.cmake:
```diff
@@ -0,0 +1,11 @@
+libdir=@CMAKE_INSTALL_FULL_LIBDIR@
+includedir=@CMAKE_INSTALL_FULL_INCLUDEDIR@
+OpenEXR_includedir=@CMAKE_INSTALL_FULL_INCLUDEDIR@/OpenEXR
```
you got the drift

dracwyrm commented:
@lgritz I updated this to the current develop branch. There is just a few things I need to check on before this can be merged. I really hope we can schedule this for the 4.3 release.

@SoapGentoo I just updated this, would you be able to look this over? There are a few things different now, like one master CMakeLists.txt file in the root to choose the packages being compiled. This means there are some variables that will point to IlmBase inside the build directory, namely these ones:
```
 ~/Storage/github/openexr $ grep -nrw . -e "OPENEXR_PACKAGE_PREFIX"
./OpenEXR/IlmImf/CMakeLists.txt:6:  SET(RUNTIME_DIR ${OPENEXR_PACKAGE_PREFIX}/bin)
./OpenEXR/IlmImf/CMakeLists.txt:9:  SET(RUNTIME_DIR ${OPENEXR_PACKAGE_PREFIX}/lib)
./OpenEXR_Viewers/CMakeLists.txt:34:INCLUDE_DIRECTORIES ( ${OPENEXR_PACKAGE_PREFIX}/include/OpenEXR )
./OpenEXR_Viewers/CMakeLists.txt:35:LINK_DIRECTORIES ( ${OPENEXR_PACKAGE_PREFIX}/lib )
./OpenEXR_Viewers/CMakeLists.txt:36:MESSAGE ( "OPENEXR_PACKAGE_PREFIX = " ${OPENEXR_PACKAGE_PREFIX})

 ~/Storage/github/openexr $ grep -nrw . -e "ILMBASE_PACKAGE_PREFIX"
./OpenEXR/CMakeLists.txt:9:IF (NOT ILMBASE_PACKAGE_PREFIX)
./OpenEXR/CMakeLists.txt:10:  SET (ILMBASE_PACKAGE_PREFIX ${CMAKE_INSTALL_PREFIX})
./OpenEXR/CMakeLists.txt:40:LINK_DIRECTORIES ( ${ILMBASE_PACKAGE_PREFIX}/lib )
./OpenEXR/CMakeLists.txt:41:MESSAGE ( "ILMBASE_PACKAGE_PREFIX = " ${ILMBASE_PACKAGE_PREFIX})
./PyIlmBase/CMakeLists.txt:14:LINK_DIRECTORIES ( ${ILMBASE_PACKAGE_PREFIX}/lib )
./PyIlmBase/CMakeLists.txt:15:MESSAGE ( "ILMBASE_PACKAGE_PREFIX = " ${ILMBASE_PACKAGE_PREFIX})
./OpenEXR_Viewers/CMakeLists.txt:9:IF (NOT ILMBASE_PACKAGE_PREFIX)
./OpenEXR_Viewers/CMakeLists.txt:10:  SET (ILMBASE_PACKAGE_PREFIX ${CMAKE_INSTALL_PREFIX})
```
I have no idea how to handle these, because if the variable is not set, then it will go to the system, but then it will be looking in /lib instead of /lib64, for example. Would you be able to advise on how to handle this properly with GNUInstallDirs? Thanks.

SoapGentoo commented:
Part of the problem here is that people want to try and use CMake with bundled dependencies and with system dependencies. This is messy and hard to get right with CMake, and I suggest relying solely on system packages.

The better, more long-term approach would be to using something like Meson, which supports system libraries and bundling as a first-class feature.

cary-ilm commented:
Just to clarify, are you saying you're getting errors in building the
current develop branch? It's still in flux at the moment, there are a
number of known problems.

On Thu, Aug 9, 2018 at 9:27 AM Jonathan Scruggs <notifications@github.com>
wrote:

> @lgritz <https://github.com/lgritz> and @cary-ilm
> <https://github.com/cary-ilm>
> I updated this to the current develop branch. There is just a few things I
> need to check on before this can be merged. I really hope we can schedule
> this for the 4.3 release.
>
> @SoapGentoo <https://github.com/SoapGentoo> I just updated this, would
> you be able to look this over? There are a few things different now, like
> one master CMakeLists.txt file in the root to choose the packages being
> compiled. This means there are some variables that will point to IlmBase
> inside the build directory, namely these ones:
>
>  ~/Storage/github/openexr $ grep -nrw . -e "OPENEXR_PACKAGE_PREFIX"
> ./OpenEXR/IlmImf/CMakeLists.txt:6:  SET(RUNTIME_DIR ${OPENEXR_PACKAGE_PREFIX}/bin)
> ./OpenEXR/IlmImf/CMakeLists.txt:9:  SET(RUNTIME_DIR ${OPENEXR_PACKAGE_PREFIX}/lib)
> ./OpenEXR_Viewers/CMakeLists.txt:34:INCLUDE_DIRECTORIES ( ${OPENEXR_PACKAGE_PREFIX}/include/OpenEXR )
> ./OpenEXR_Viewers/CMakeLists.txt:35:LINK_DIRECTORIES ( ${OPENEXR_PACKAGE_PREFIX}/lib )
> ./OpenEXR_Viewers/CMakeLists.txt:36:MESSAGE ( "OPENEXR_PACKAGE_PREFIX = " ${OPENEXR_PACKAGE_PREFIX})
>
>  ~/Storage/github/openexr $ grep -nrw . -e "ILMBASE_PACKAGE_PREFIX"
> ./OpenEXR/CMakeLists.txt:9:IF (NOT ILMBASE_PACKAGE_PREFIX)
> ./OpenEXR/CMakeLists.txt:10:  SET (ILMBASE_PACKAGE_PREFIX ${CMAKE_INSTALL_PREFIX})
> ./OpenEXR/CMakeLists.txt:40:LINK_DIRECTORIES ( ${ILMBASE_PACKAGE_PREFIX}/lib )
> ./OpenEXR/CMakeLists.txt:41:MESSAGE ( "ILMBASE_PACKAGE_PREFIX = " ${ILMBASE_PACKAGE_PREFIX})
> ./PyIlmBase/CMakeLists.txt:14:LINK_DIRECTORIES ( ${ILMBASE_PACKAGE_PREFIX}/lib )
> ./PyIlmBase/CMakeLists.txt:15:MESSAGE ( "ILMBASE_PACKAGE_PREFIX = " ${ILMBASE_PACKAGE_PREFIX})
> ./OpenEXR_Viewers/CMakeLists.txt:9:IF (NOT ILMBASE_PACKAGE_PREFIX)
> ./OpenEXR_Viewers/CMakeLists.txt:10:  SET (ILMBASE_PACKAGE_PREFIX ${CMAKE_INSTALL_PREFIX})
>
> I have no idea how to handle these, because if the variable is not set,
> then it will go to the system, but then it will be looking in /lib instead
> of /lib64, for example. Would you be able to advise on how to handle this
> properly with GNUInstallDirs? Thanks.
>
> @meshula <https://github.com/meshula> Just updated this, if you are
> interested in it. :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openexr/openexr/pull/242#issuecomment-411816621>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AUWxmcCKevnokB7VucJFdI9A8TFVX2_Wks5uPGLegaJpZM4O0fEU>
> .
>


-- 
Cary Phillips | R&D Supervisor | ILM San Francisco

dracwyrm commented:
@cary-ilm Just the usual developer errors. This patch changes hardcoded directories to GNUInstallDirs, which are a standard. For example, Gentoo installs 64 bit libraries to /usr/lib64, but the directories are hardcoded to just lib, thus making installation harder. GNUInstallDirs are known by packages for distros, so directories can easily be changed to whatever the distro needs, and this works for Windows and Mac. This patch then adds in the 3 missing Pkg-Config files with CMake builds.

The main thing I need help with that I asked Soap for, is the files listed in the code block as I don't know how to handle them properly in changing over to GNUInstallDirs.

I hope this is more clear. I have a major headache today as I got into a lot of dust at work that I'm allergic to.

meshula commented:
I think the use of GNUInstallDirs makes sense if all the variables are relative to CMAKE_INSTALL_PREFIX. The documentation ----  https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html --- is ambiguous. Are the variables relative (lib, doc), respecting the CMAKE_INSTALL_PREFIX variable, or do they impose on linux /usr paths, and on windows c:/Program Files/ paths?

dracwyrm commented:
> Finally, I think the use of the PACKAGE_PREFIX variables would be vestigial if the GNUInstallDirs works out, i.e. lines referencing the ILMBASE/OPENEXR_PACKAGE_PREFIX would have to be updated to use those CMAKE_INSTALL variables consistently.

I was thinking of using if statements for the *_PACKAGE_PREFIX variables. The modules of OpenEXR can be built individually, so in that case, you need to look at the system for the include dirs, but if you are building everything in one go, then those variables need to be set to the build directory. Though, it should be possible to build everything separately, but currently, if you want the viewers, then you need to compile it with OpenEXR instead of it using system resources, so I might need to fix that. On Gentoo, we have four different ebuilds for each module.

SoapGentoo commented:
@meshula *most* of GNUInstallDirs is relative to `CMAKE_INSTALL_PREFIX` (there are some exceptions, especially those dealing with `/etc`-like directories: https://github.com/Kitware/CMake/blob/master/Modules/GNUInstallDirs.cmake#L80).

In general, I would rather hijack `CMAKE_INSTALL_LIBDIR` if on Windows, set it to the same value as `CMAKE_INSTALL_BINDIR` before including `GNUInstallDirs`, and then replace all occurrences of `RUNTIME_DIR` with `CMAKE_INSTALL_BINDIR`. Reducing the number of own variables for upstream/Kitware defined ones greatly reduces cognitive load, without losing anything in practice here.

SoapGentoo commented on IlmBase/CMakeLists.txt:
```diff
@@ -150,11 +150,11 @@ FILE ( REMOVE ${ILMBASE_TMP_CONFIG} )
+  CREATE_LIBTOOL_FILE ( IexMath /${CMAKE_INSTALL_LIBDIR} )
+  CREATE_LIBTOOL_FILE ( Imath /${CMAKE_INSTALL_LIBDIR} )
+  CREATE_LIBTOOL_FILE ( IlmThread /${CMAKE_INSTALL_LIBDIR} )
 ENDIF ()
```
remove all of this. Creating libtool archives in CMake is a cardinal sin. Use pkgconfig and never, **ever** use Libtool archives. They're awful, clumsy, hard to use and CMake has no support for them.

SoapGentoo commented on IlmBase/CMakeLists.txt:
```diff
@@ -170,26 +170,10 @@ ENDIF ()
-    lib/pkgconfig
-  )
-ENDIF()
+if ( PKG_CONFIG_FOUND )
```
change this to `if ( NOT WIN32 )` or something. Making this conditional on whether you found `pkg-config` is brittle.

SoapGentoo commented on OpenEXR/CMakeLists.txt:
```diff
@@ -230,5 +230,10 @@ INSTALL ( FILES
+  ${CMAKE_INSTALL_DOCDIR}/examples
   )
+
+if ( PKG_CONFIG_FOUND )
```
ditto

SoapGentoo commented on OpenEXR_Viewers/CMakeLists.txt:
```diff
@@ -57,6 +57,10 @@ ENDIF ()
+  ${CMAKE_INSTALL_DOCDIR}
   )
 
+if ( PKG_CONFIG_FOUND )
```
ditto

SoapGentoo commented on PyIlmBase/CMakeLists.txt:
```diff
@@ -49,3 +49,7 @@ IF (ENABLE_TESTS)
   ADD_SUBDIRECTORY ( PyImathTest )
 ENDIF ()
 
+if ( PKG_CONFIG_FOUND )
```
ditto

dracwyrm commented:
I guess this didn't make it into v2.3 release. :(

dracwyrm commented:
@cary-ilm I guess this make it in v2.3.0 release. The CMake build system won't have Pkg-Config files for 3 modules: OpenEXR, PyIlmBase, and OpenEXR_Viewers. The OpenEXR module is the important one as not having this will break any package using Pkg-Config to find the libraries and include dirs. This is not good. I was hoping we could have gotten this PR in to fix this because this is the next minor release, which is more appropriate for these types of changes than a point release. I can create a PR with just Pkg-Config files and no GNUInstallDirs. Is there a way of convincing you to pull the current 2.3 release to add that in or make a 2.3.1 release to fix the missing Pkg-Config file issue?

dracwyrm commented:
@cary-ilm Thanks for replying back. I think there is a slight thing I need to fix before this is merged. I realize these are more radical changes, which is why I made the other PR with just .pc files, if you want to do more testing on this PR first. :)

cary-ilm commented:
Superseded by #332, #341
